diagnostics/buffer.go

```diff
@@ -0,0 +1,124 @@
+package diagnostics
+
+import (
+	"sync/atomic"
+
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/topdown"
+)
+
+var seq uint64
+
+// Buffer stores diagnostic information. They are not safe for concurrent access.
+type Buffer struct {
+	ring []*Info
+
+	size int
+
+	start int // The index of the next item to pop.
+	end   int // The index where the next item will be inserted.
+}
+
+// NewBuffer creates a new Buffer with maximum size n. NewBuffer will panic if n is not
+// positive.
+func NewBuffer(n int) *Buffer {
+	if n <= 0 {
+		panic("size must be greater than 0")
+	}
+
+	return &Buffer{ring: make([]*Info, n, n)}
+}
+
+// Push inserts the provided Info into the Buffer. If the Buffer is full, the oldest item
+// is deleted to make room.
+func (b *Buffer) Push(info *Info) {
+	b.ring[b.end] = info
+
+	b.end = (b.end + 1) % len(b.ring)
+
+	switch b.size {
+	case len(b.ring): // We erased something, move the start up.
+		b.start = (b.start + 1) % len(b.ring)
+	default:
+		b.size++
+	}
+}
+
+// Pop removes the oldest item from the Buffer and returns it. If the Buffer is empty, ok
+// will be false.
+func (b *Buffer) Pop() (info *Info, ok bool) {
+	if b.size == 0 {
+		return nil, false
+	}
+
+	info = b.ring[b.start]
+	b.ring[b.start] = nil // Free it up for garbage collection.
+
+	b.start = (b.start + 1) % len(b.ring)
+	b.size--
+
+	return info, true
+}
+
+// Iter iterates over the buffer. If fn returns true, iteration stops and Iter returns
+// true. Otherwise, iteration completes and Iter returns false.
+func (b *Buffer) Iter(fn func(*Info) bool) bool {
+	i := b.start
+	for j := 0; j < b.size; j++ {
+		if fn(b.ring[i]) {
+			return true
+		}
+
+		i = (i + 1) % len(b.ring)
+	}
+	return false
+}
+
+// Info stores diagnostic information about the evaluation of a query.
+type Info struct {
+	ID uint64
+
+	Query   string
+	Results interface{}
+
+	Metrics metrics.Metrics
+	Trace   *topdown.BufferTracer
```

Comment:
The only reason I don't like that is then we have to do a type assertion to actually extract the trace, and I didn't see anywhere in the server where we use something other than a BufferTracer